apps/vscode-wing/src/console/console-manager.ts

```diff
@@ -0,0 +1,274 @@
+        })
+      );
+      await activePanel.client.runTest(test.id);
+    });
+
+    commands.registerCommand("wingConsole.runAllTests", async () => {
+      if (!activeInstanceId) {
+        return;
+      }
+      const activePanel = getInstance(activeInstanceId);
+      if (!activePanel) {
+        return;
+      }
+      const tests = testsExplorer.getTests();
+      testsExplorer.update(
+        tests.map((testItem) => {
+          return {
+            ...testItem,
+            status: "running",
+          };
+        })
+      );
+      await activePanel.client.runAllTests();
+    });
+  };
+
+  const updateLogs = async (instance: ConsoleInstance) => {
+    const logs = await instance.client.getLogs({ time: logsTimestamp });
+
+    logs.forEach((log) => {
+      logger.appendLine(`[${log.level}] ${log.message}`);
+    });
+    logsTimestamp = Date.now();
+  };
+
+  const getTheme = () => {
+    const theme = window.activeColorTheme.kind;
+    if (theme === 1) {
+      return "light";
+    }
+    return "dark";
+  };
+
+  const getInstance = (instanceId: string) => {
+    if (!instanceId) {
+      return;
+    }
+    return instances[instanceId];
+  };
+
+  const addInstance = async (instance: ConsoleInstance) => {
+    logger.appendLine(`Wing Console is running at http://${instance.url}`);
+
+    instance.client.onInvalidateQuery({
+      onData: async (key) => {
+        if (activeInstanceId !== instance.id) {
+          return;
+        }
+        if (timeout) {
+          clearTimeout(timeout);
+        }
+        timeout = setTimeout(async () => {
+          if (!key || key === "app.logs") {
+            await updateLogs(instance);
+          }
+          resourcesExplorer.update(await instance.client.listResources());
+          testsExplorer.update(await instance.client.listTests());
+        }, 300);
+      },
+      onError: (err) => {
+        logger.appendLine(err);
+      },
+    });
+    instances[instance.id] = instance;
+
+    await setActiveInstance(instance.id);
+  };
+
+  const setActiveInstance = async (instanceId: string) => {
+    const instance = getInstance(instanceId);
+    if (!instance) {
+      return;
+    }
+
+    if (!webviewPanel) {
+      webviewPanel = window.createWebviewPanel(
+        VIEW_TYPE_CONSOLE,
+        `Console`,
+        ViewColumn.Beside,
+        {
+          enableScripts: true,
+          enableCommandUris: true,
+        }
+      );
+
+      webviewPanel.iconPath = {
+        light: Uri.joinPath(
+          context.extensionUri,
+          "resources",
+          "icon-light.png"
+        ),
+        dark: Uri.joinPath(context.extensionUri, "resources", "icon-dark.png"),
+      };
+
+      webviewPanel.onDidDispose(async () => {
+        resourcesExplorer.clear();
+        testsExplorer.clear();
+        webviewPanel = undefined;
+        activeInstanceId = undefined;
+
+        Object.keys(instances).forEach(async (id) => {
+          await closeInstance(id);
```

Comment:
it results an unexpected behavior in my opinion:
let's say I have two `.w` files: `file1.w` and `file2.w`.
I opened the console for both and now i'm switching between one console view to another.
I now finished all my testings for `file2.w` and I close the console view for it (clicking X in the console tab).
I expect to see the console view for `file1.w` which i'm still working on while In reality, all console instances were closed.
In order to see the console for `file1.w` I need to click on the (W) button again to run the console extension

Reply:
I think it would be weird to reopen the console view if you just closed the tab. Also, this behavior will depend on whether you opened a file while having the console opened (otherwise the file won't have an "instance" in the console).

If we do so, would you close the focused Editor as well when closing the Console tab? If not, should we focus on the other opened wing file and then, open the Console tab again?